### PR TITLE
Simplicity

### PR DESCRIPTION
**Pull request by roconnor-blockstream**
Tasks needed to pass CI:

- [x] For Mingw32, remove pthread dependency from Simplicity (upstream).
- [x] Merge (upcoming) libsecp256k1 release into Simplicity (upstream).

Tasks needed for public signet:
- [x] Add weight requirements for jets and combinantors (upstream).
- [x] Maximum memory allocation (upstream).
- [x] Other missing anti-DOS protection (upstream)?

Tasks needed for testing.
- [ ] Ensure that the fuzzer covers `libelementssimplicity`.

Before merging:
- [ ] Rebase on master.

### CODE REVIEW

**Comment by delta1** [comment]
changed the base branch to 23.x since this was rebased @roconnor-blockstream

**Comment by apoelstra** [comment]
In 10ff078991e4565c8f74fe9287063124f42b96ec you make simplicity a subtree (it appears not to be before then).

You should also update `ci/lint/06_script.sh` to add a check that the subtree has been correctly updated. (Currently passes in the commit where you add the subtree, but fails after the next commit 002c00949995ba60b672d5ce97549b4a42cc9523 where you illegally add a sources.mk file to what should be a subtree directory. It appears that other cases where there is a sources.mk file, this exists upstream.)

**Comment by apoelstra** [comment]
CI failures look real. e.g. on the ASan build it originates in
```
2024-10-02T15:23:10.952000Z TestFramework (ERROR): Assertion failed
4069 Traceback (most recent call last):
4070   File "/tmp/cirrus-ci-build/ci/scratch/build/elements-x86_64-pc-linux-gnu/test/functional/test_framework/test_framework.py", line 132, in main
4071     self.run_test()
4072   File "/tmp/cirrus-ci-build/ci/scratch/build/elements-x86_64-pc-linux-gnu/test/functional/feature_taproot.py", line 1820, in run_test
4073     self.test_spenders(self.nodes[1], spenders_taproot_active(), input_counts=[1, 2, 2, 2, 2, 3])
4074   File "/tmp/cirrus-ci-build/ci/scratch/build/elements-x86_64-pc-linux-gnu/test/functional/feature_taproot.py", line 1552, in test_spenders
4075     self.block_submit(node, [tx], msg, witness=True, accept=fail_input is None, cb_pubkey=cb_pubkey, fees=fee, sigops_weight=sigops_weight, err_msg=expected_fail_msg)
4076   File "/tmp/cirrus-ci-build/ci/scratch/build/elements-x86_64-pc-linux-gnu/test/functional/feature_taproot.py", line 1331, in block_submit
4077     assert block_response is not None and err_msg in block_response, "Missing error message '%s' from block response '%s': %s" % (err_msg, "(None)" if block_response is None else block_response, msg)
4078 AssertionError: Missing error message 'Invalid Schnorr signature hash type' from block response 'non-mandatory-script-verify-flag (Program's CMR does not match)': simplicity/empty_program,sighash/keypath_unk_hashtype_7*
4079 2024-10-02T15:23:11.007000Z TestFramework (INFO): Stopping nodes
```
in feature_taproot.py

File: test/functional/feature_taproot.py [hunk 1]
```diff
@@ -1161,6 +1170,13 @@ def predict_sigops_ratio(n, dummy_size):
         tap = taproot_construct(pubs[0], scripts)
         add_spender(spenders, "alwaysvalid/notsuccessx", tap=tap, leaf="op_success", inputs=[], standard=False, failure={"leaf": "normal"}) # err_msg differs based on opcode
 
+    # == Simplicity tests ==
+
+    tap = taproot_construct(pubs[0], [("simplicity_iden", bytes.fromhex("dbfefcfc7796acfc86b435c1f81ed8a165dab2649dc48b0f35f832647868fb5e"), LEAF_VERSION_TAPSIMPLICITY)])
```
Comment (roconnor-blockstream) [inline test/functional/feature_taproot.py:1175]: CMR needs to be updated.

**Comment by apoelstra** [comment]
code review ACK 0aeafeb0d6e8d06d65f797ce0f55748a48e84e0b

**Comment by apoelstra** [comment]
4 CI failures -- the windows one might be real (`test_bitcoin` failing); the TSan one looks like maybe we need new tsan whitelist entries; the other two are just environment/outdated CI failures.

**Review by delta1 — APPROVED** [review]
utACK 508836028e64f68dc5a8c068cdbf14080de1479c 

pending CI at https://cirrus-ci.com/build/5064884356907008

**Review by apoelstra — APPROVED** [review]
utACK 508836028e64f68dc5a8c068cdbf14080de1479c

File: Makefile.am [hunk 2]
```diff
@@ -195,6 +195,7 @@ LCOV_FILTER_PATTERN = \
 	-p "src/crypto/ctaes" \
 	-p "src/minisketch" \
 	-p "src/secp256k1" \
+	-p "src/simplicity/secp256k1" \
```
Comment (apoelstra) [inline Makefile.am:198]: In 50916cf1081d842865611130cb51dda38163cc95:

Probably we didn't want to filter this. (I'm kinda surprised that the upstream libsecp is filtered. But I guess they have their own coverage analysis.)

Comment (apoelstra) [reply Makefile.am:198]: While we're at it, we should stick /nix/store into this list and src/include/boost. I'm not sure why we need to do this and Bitcoin doesn't (well, the answer is "NixOS") but it's harmless and helpful for Russell and me.

Comment (roconnor-blockstream) [reply Makefile.am:198]: I'm a little hesitant to add nix-specific fixes here, and rather just patch the file in https://github.com/roconnor-blockstream/elements-nix/

Comment (roconnor-blockstream) [reply Makefile.am:198]: I've removed `src/simplicity/secp256k1` from the filter as part of #1390.

**Pull request merged**